core/headerchain.go

```diff
@@ -105,302 +122,212 @@ func NewHeaderChain(chainDb ethdb.Database, config *params.ChainConfig, engine c
+			return hc.GetHeaderByHash(canonicalHash)
 		}
+		header = hc.GetHeader(header.ParentHash[types.QuaiNetworkContext], header.Number64()-1)
 	}
 
-	return 0, nil
 }
 
-// InsertHeaderChain inserts the given headers.
-//
-// The validity of the headers is NOT CHECKED by this method, i.e. they need to be
-// validated by ValidateHeaderChain before calling InsertHeaderChain.
-//
-// This insert is all-or-nothing. If this returns an error, no headers were written,
-// otherwise they were all processed successfully.
-//
-// The returned 'write status' says if the inserted headers are part of the canonical chain
-// or a side chain.
-func (hc *HeaderChain) InsertHeaderChain(chain []*types.Header, start time.Time) (WriteStatus, error) {
-	if hc.procInterrupt() {
-		return 0, errors.New("aborted")
-	}
-	res, err := hc.writeHeaders(chain)
+// loadLastState loads the last known chain state from the database. This method
+// assumes that the chain manager mutex is held.
+func (hc *HeaderChain) loadLastState() error {
+	// TODO: create function to find highest block number and fill Head FIFO
+	headsHashes := rawdb.ReadHeadsHashes(hc.headerDb)
 
-	// Report some public statistics so the user has a clue what's going on
-	context := []interface{}{
-		"count", res.imported,
-		"elapsed", common.PrettyDuration(time.Since(start)),
+	heads := make([]*types.Header, maxHeadsQueueLimit)
+	for i, hash := range headsHashes {
+		heads[i] = hc.GetHeaderByHash(hash)
 	}
-	if err != nil {
-		context = append(context, "err", err)
-	}
-	if last := res.lastHeader; last != nil {
-		context = append(context, "number", last.Number, "hash", res.lastHash)
-		if timestamp := time.Unix(int64(last.Time), 0); time.Since(timestamp) > time.Minute {
-			context = append(context, []interface{}{"age", common.PrettyAge(timestamp)}...)
-		}
+	hc.heads = heads
+
+	return nil
+}
+
+// Blockchain retrieves the blockchain from the headerchain.
+func (hc *HeaderChain) BlockChain() *BlockChain {
+	return hc.bc
+}
+
+// NOTES: Headerchain needs to have head
+// Singleton Tds need to get calculated by slice after successful append and then written into headerchain
+// Slice uses HLCR to query Headerchains for Tds
+// Slice is a collection of references headerchains
+
+// GetBlockNumber retrieves the block number belonging to the given hash
+// from the cache or database
+func (hc *HeaderChain) GetBlockNumber(hash common.Hash) *uint64 {
+	if cached, ok := hc.numberCache.Get(hash); ok {
+		number := cached.(uint64)
+		return &number
 	}
-	if res.ignored > 0 {
-		context = append(context, []interface{}{"ignored", res.ignored}...)
+	number := rawdb.ReadHeaderNumber(hc.headerDb, hash)
+	if number != nil {
+		hc.numberCache.Add(hash, *number)
 	}
-	log.Info("Imported new block headers", context...)
-	return res.status, err
+	return number
 }
 
 // GetBlockHashesFromHash retrieves a number of block hashes starting at a given
```

Comment:
Please consider renaming `GetBlockHashesFromHash()` to `GetAncestors()` or `GetAncestorHashes()` or something like that. The current name is not intuitive IMO